lib/Sema/TypeCheckDecl.cpp

```diff
@@ -3151,25 +3155,26 @@ class DeclChecker : public DeclVisitor<DeclChecker> {
 
     checkUnsupportedNestedType(PD);
 
+    // Check for circular inheritance within the protocol.
+    SmallVector<ProtocolDecl *, 8> path;
+    path.push_back(PD);
+    checkCircularity(TC, PD, diag::circular_protocol_def,
+                     DescriptiveDeclKind::Protocol, path);
+    // Skip full validation if circularity exists.
+    if (PD->isInvalid())
```

Comment:
Someone else can still call validateDecl() on the protocol, though, for example if a name lookup references a protocol in another file. This is not really the right fix unfortunately.